cql3/query_processor.cc

```diff
@@ -1084,6 +1084,9 @@ void query_processor::migration_subscriber::on_update_aggregate(const sstring& k
 void query_processor::migration_subscriber::on_update_view(
         const sstring& ks_name,
         const sstring& view_name, bool columns_changed) {
+    // scylladb/scylladb#16392 - Materialized views are also tables so we need at least handle
+    // them as such when changed.
+    on_update_column_family(ks_name, view_name, columns_changed);
```

Comment:
Very nice catch.

I want to point out to whoever is the expert in these "on_*" things that they are very fragile, poorly documented and very easy to misuse. About a month ago in 88a5ddabcec625434b2df0ed82a63c36245af35f I had to correct a very similar bug with creation of materialized views - these didn't get tablets created for them because the tablet creation was moved into a `on_before_create_column_family()` callback, and this wasn't called for new views, even though they are column families too.

In my opinion, these callbacks sound on paper like they are a good idea because it allows "mixing" different concerns easily - e.g., the code that deals with creating a table doesn't need to know about the existence of tablets. But in reality, the developers don't know about all these invisible assumptions and often make bugs - the bug that Eliran found here probably existed for at least 7 years (maybe we even copied this design from Cassandra?)

@tgrabiec you may be interested in this rant :-)

Reply:
What's the alternative?

Reply:
The alternative is more "monolithic" code - you would have a "create table" code which also create the tablets, and since creating a view **also** creates a table (it doesn't duplicate the entire table creation code), it would naturally call the table-creation function and everything related to creating tables - including creating their tablets - would be done.
Anyway, I'm not saying I have the perfect solution. Just that it's the second bug we solved in a month that ended up being table-related callbacks which we forgot to call for the case of views (which, as Eliran noted, are also tables).

Reply:
I think it's unrealistic to glue everything directly, and doesn't guarantee we don't forget to add the glue.